src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerProfileUpdateService.kt

```diff
@@ -44,7 +51,61 @@ class PrisonerProfileUpdateService(
     }
   }
 
-  private fun Optional<Offender>.orElseThrowNotFound(message: String, prisonerNumber: String) =
+  @Transactional
+  fun updateNationalityOfCurrentBooking(prisonerNumber: String, nationality: String?) {
+    val profileType: ProfileType = profileTypeRepository.nationalityProfile().getOrThrow()
+    val profileCode: ProfileCode? =
+      nationality?.uppercase()?.let { profileCodeRepository.profile(profileType, it).getOrThrow() }
+
+    try {
+      val currentBooking = currentBooking(prisonerNumber)
+      val currentNationality = currentBooking.profileDetails
+        ?.firstOrNull { it.code?.id?.type?.equals(profileType) ?: false }
+
+      if (nationality == null && currentNationality != null) {
+        currentBooking.profileDetails.remove(currentNationality)
+      } else {
+        if (currentNationality != null) {
+          currentNationality.setProfileCode(profileCode)
+        } else {
+          currentBooking.profileDetails.add(
+            OffenderProfileDetail.builder()
+              .id(OffenderProfileDetail.PK(currentBooking, profileType, 1))
+              .caseloadType("INST")
+              .code(profileCode)
+              .listSequence(profileType.listSequence)
+              .build(),
+          )
+        }
+      }
+    } catch (e: CannotAcquireLockException) {
+      throw processLockError(e, prisonerNumber)
+    }
+  }
+
+  private fun currentBooking(prisonerNumber: String) =
+    (
+      offenderRepository.findLinkedToLatestBookingForUpdate(prisonerNumber)
```

Comment:
As discussed, I think we need to do the locking on the actual OFFENDER_PROFILE_DETAILS (NAT) record so that we get the exception thrown if someone's trying to edit nationality in NOMIS at the same time.  I think this should do it: https://github.com/ministryofjustice/prison-api/pull/2313